LG-4143: Include conversation ID in LexisNexis exceptions

aduth opened the pull request:
**Why**: So that we can query or associate exception logging events based on transaction ID (and reference after #5194 is merged).

Draft, currently thinking about...

- Should `Proofing::Base#proof` be made to behave similar to changes here to retain most of the `result`?
   - i.e. change this:
      - https://github.com/18F/identity-idp/blob/3220ddef03883a6884365456366ab96b6a2cf3bb/app/services/proofing/base.rb#L43-L48
   - to...
   ```rb
      result = Proofing::Result.new 
      begin
        execute_proof(proofer, vendor_applicant, result) 
      rescue => exception
        result.exception = exception
      end
      result
    end
   ``` 
   - Would this allow `LexisNexis::Proofer.proof_applicant` to not have to `rescue`?
- Can we consolidate how `transaction_id` is assigned to the result, to avoid having it between the main logic path and the `rescue` block?
   - e.g. Can we pass `result` into `send_verification_request` and assign to it before raising any errors?
- Could we avoid raising errors altogether for these known and well-structured failures?
   - e.g. Treat them more like traditional result errors than exceptions

amathews-fs commented:
For your last point: I vote strongly to remove the tracking_ids from the message (I know I added them). The data is now included elsewhere in the JSON log payload and it causes problems aggregating exceptions because with the tracking every exception is unique.

aduth commented:
I pushed some updates to account for the merge of #5194, and to follow the question originally posed in the pull request comment:

>- Could we avoid raising errors altogether for these known and well-structured failures?
>   - e.g. Treat them more like traditional result errors than exceptions

The commit 61b7fde includes this refactor.

> For your last point: I vote strongly to remove the tracking_ids from the message (I know I added them). The data is now included elsewhere in the JSON log payload and it causes problems aggregating exceptions because with the tracking every exception is unique.

Incorporated this in 209fef7, further expanded to cover other base error message tracking IDs in 61b7fde.

zachmargolis commented:
> Should `Proofing::Base#proof` be made to behave similar to changes here to retain most of the `result`?

I really dislike `Proofing::Base#proof`:
1. it swallows all exceptions
2. its interface is really abstract and doesn't make it clear what should be passed

We've moved beyond `#proof` as a generic interface to all proofers (because now we have separate steps for separate proofers) I would be in favor of removing it entirely for specific method for specific steps to make it easier to tell what's going on

zachmargolis commented on spec/jobs/resolution_proofing_job_spec.rb:
```diff
@@ -152,7 +152,7 @@
           {
             'Status' => {
-              'CoversationId' => lexisnexis_transaction_id,
+              'ConversationId' => lexisnexis_transaction_id,
```
😮 yikes, great catch

zachmargolis commented on app/services/proofing/lexis_nexis/verification_error_parser.rb:
```diff
@@ -38,19 +38,38 @@ def verification_status
+      end
+
+      def conversation_id
+        body.dig('Status', 'ConversationId')
```
code coverage seems to think this isn't covered?

aduth replied:
> code coverage seems to think this isn't covered?

Ah, it's unused now. Prior to 61b7fdef72c1903023d89437d8cce626a3240960 we were embedding the conversation ID in the error message.

zachmargolis replied:
Should we remove it? Or do we still use it in success cases?

aduth replied:
I'll remove. The conversation ID is still used, but a step up in the response class, which is relevant to the intended change here since the goal is to make sure that we include the conversation ID even in these instances where previously we'd have raised errors on unexpected status codes.

https://github.com/18F/identity-idp/blob/2f7c0ee15cdca484d762de6c9d4a76bef7c1f3aa/app/services/proofing/lexis_nexis/response.rb#L36-L38

aduth commented:
> > Should `Proofing::Base#proof` be made to behave similar to changes here to retain most of the `result`?
> 
> I really dislike `Proofing::Base#proof`:
> 
> 1. it swallows all exceptions
> 2. its interface is really abstract and doesn't make it clear what should be passed
> 
> We've moved beyond `#proof` as a generic interface to all proofers (because now we have separate steps for separate proofers) I would be in favor of removing it entirely for specific method for specific steps to make it easier to tell what's going on

For context, the initial implementation added a [`rescue` block](https://github.com/18F/identity-idp/pull/5198/commits/61b7fdef72c1903023d89437d8cce626a3240960#diff-77f8148755be41aa5a88db9d0ab4e44c7a0b2742b4c0186b06de583d8766ab92L40) to `Proofing::LexisNexis::Proofer`, and the quoted text was contemplating this idea of being more accommodating to raised errors at least in letting a result maintain any properties it has, vs. how it exists today where we [create a new, empty result](https://github.com/18F/identity-idp/blob/2f7c0ee15cdca484d762de6c9d4a76bef7c1f3aa/app/services/proofing/base.rb#L47).

The rescue block was removed in 61b7fdef72c1903023d89437d8cce626a3240960. And to your first point, two errors that we'd previously raised and swallowed rather quietly are now handled more directly within the error generator as like with other validation errors, instead of as an "exception".

Overall then, I think this helps move in the direction you're suggesting? Fewer raised errors, and not further entrenching ourselves to the proofer class with enhancements to the vendor-specific proofer implementations.

zachmargolis commented:
> Overall then, I think this helps move in the direction you're suggesting? Fewer raised errors, and not further entrenching ourselves to the proofer class with enhancements to the vendor-specific proofer implementations.

Yup! Sorry didn't mean to detract from this PR, I just found a soap box and jumped on it

zachmargolis reviewed (approved):
LGTM

amathews-fs reviewed (approved):
LGTM! Let's make sure to exercise this on staging when it gets there.
